### ISSUE TITLE
" failed to rename compilation results " Zig build fails to compile while `zig build run` is active.

### ISSUE THREAD

**Issue by MassKlaus** [ziglang/zig#21168]
### Zig Version

0.13.0

### Steps to Reproduce and Observed Behavior

I have a hot reload plugin system added to my project on windows. While the main project is running attempting to recompile the shared library results in the zig.build compilation to fail and throw this error.

```
error: failed to rename compilation results ('C:\...\.zig-cache\tmp\59219da4de8d4725') into local cache ('C:\...\.zig-cache\o\dbf31f96302da5d52610317112474912'): AccessDenied
```

Modifying the zig.build by adding a comment line results in the library compiling with 0 issues.

Here is the repo used for this test: [Repo](https://github.com/MassKlaus/syalizer)

### Expected Behavior

`zig build -Dplug_only=true` To compile normally.

**Comment by rohlem** [comment]
I'd assume that a `.dll` file currently loaded into and used by a process cannot be overwritten. That would explain an `AccessDenied` error.
Since `zig build run` probably runs your main program, do you know / have you checked whether the `.dll` file is unloaded by the program at the point you trigger the next compilation?

If I'm not wrong about this being the issue, then I'm not sure whether the build system integration can/should do much to work around such a system-level limitation.
Maybe a userland workaround could be that the program always copies/moves the dll file to load it under a different name/path, then monitors the original dll file for changes/rebuilds and redoes this process every time to reload the file for every new change?

**Comment by MassKlaus** [comment]
I unload the `.dll` manually before starting the compilation but even for that case. I build the library as `syalizer.plug.next.dll` when the `-Dplug_only=true` flag is set.

I believe the access denied is related to the hash generated by the compiler in .zig-cache when compiling the `build.zig` file it self as just adding a new line in the middle makes the `.dll` compile with no problems and hot swaps with no issues.

**Comment by rohlem** [comment]
> just adding a new line in the middle makes the `.dll` compile with no problems and hot swaps with no issues.

I'm not sure where you're adding this newline. Do you see effects of the hot swapping within the running program? Your original post says

> Modifying the zig.build by adding a comment line results in the library compiling with 0 issues.

however the build system caches every step based on the inputs.
If the `build.zig` is modified but the compilation would be the same (same input files, flags etc.), the compilation is skipped.
From what I understand this would also skip the file copy, so wouldn't trigger an `AccessDenied` either - it would just do nothing.

> This also only happens when using `zig build run`. If I launch the program through the `.exe` directly and then attempt a compilation it goes through with no issues which I believe removes the `.dll` being loaded doubt.

I'm not sure whether the run step runs the .exe from the location it is installed.
I think the build system is perfectly capable of compiling a `.exe` file and running it even if it never installs it (copies it outside of the zig cache directory), therefore the run step might also launch it from within the cache directory.
Technically the installation step (`zig build` or `zig build install`) should fail though while the program is running, from what I understand.

**Comment by MassKlaus** [comment]
I am not knowledgeable on the inner workings of the compiler so I am not 100% but whenever I run `zig build -Dplug_only=true`

I always see this output which I just noticed: 
![Code_6fXXdWEhi1](https://github.com/user-attachments/assets/66556abe-e20d-452d-ba3b-8d5b50f4c8a9)

This will show every time I run the command

In the case I run the `.exe` manually and I make sure the `.dll` to be replaced is already busy I get this message

```
unable to update file from 'C:\...\syalizer\.zig-cache\o\8dec672f861e4b9a5a05785b20289173\syalizer.plug.next.dll' to 'C:\...\syalizer\zig-out\bin\syalizer.plug.next.dll': AccessDenied
```

> I think the build system is perfectly capable of compiling a .exe file and running it even if it never installs it

I think that is what could be happening, the compile script is held in the cache with a handle, s attempting to copy the file to the cache with the same hash ends up with an open file that is being process/used therefore resulting in an AccessDenied.

**Comment by MassKlaus** [comment]
Figured out that due to some odd caching, the build script was recompiling itself. Deleting the cache and rebuilding fixed it

**Comment by FMurray** [comment]
I know this is closed, but I had a similar error message on MacOS and it turns out that iCloud sync running on the directory also causes a similar error message.

**Comment by rohlem** [comment]
@FMurray Just FYI, https://github.com/ziglang/zig/issues/17342 seems to be the more iCloud-centric/-relevant issue .

**Comment by MassKlaus** [comment]
I've seen that issue but I am fully om Windows so I think with no iTunes or any cloud tooling actually

**Comment by KarryCharon** [comment]
@rohlem So this issuse is not fixed on mac. right?
This is my log:
```log
➜  Zig mkdir L    
➜  Zig cd L           
➜  L zig init                  
info: created build.zig
info: created build.zig.zon
info: created src/main.zig
info: created src/root.zig
info: see `zig build --help` for a menu of options
➜  L zig build -Dplug_only=True
error: failed to rename compilation results ('/Volumes/Code/Zig/L/.zig-cache/tmp/034576442839a131') into local cache ('/Volumes/Code/Zig/L/.zig-cache/o/e3d9f058575d2a681706b7facbb05072'): AccessDenied
➜  L zig build                 
error: failed to rename compilation results ('/Volumes/Code/Zig/L/.zig-cache/tmp/49a39d08c090f394') into local cache ('/Volumes/Code/Zig/L/.zig-cache/o/e3d9f058575d2a681706b7facbb05072'): AccessDenied
```

**Comment by KarryCharon** [comment]
> [@rohlem](https://github.com/rohlem) So this issuse is not fixed on mac. right? This is my log:
> 
> ```
> ➜  Zig mkdir L    
> ➜  Zig cd L           
> ➜  L zig init                  
> info: created build.zig
> info: created build.zig.zon
> info: created src/main.zig
> info: created src/root.zig
> info: see `zig build --help` for a menu of options
> ➜  L zig build -Dplug_only=True
> error: failed to rename compilation results ('/Volumes/Code/Zig/L/.zig-cache/tmp/034576442839a131') into local cache ('/Volumes/Code/Zig/L/.zig-cache/o/e3d9f058575d2a681706b7facbb05072'): AccessDenied
> ➜  L zig build                 
> error: failed to rename compilation results ('/Volumes/Code/Zig/L/.zig-cache/tmp/49a39d08c090f394') into local cache ('/Volumes/Code/Zig/L/.zig-cache/o/e3d9f058575d2a681706b7facbb05072'): AccessDenied
> ```

Additional info, this project dir is on my truenas smb volume

**Comment by KarryCharon** [comment]
> [@KarryCharon](https://github.com/KarryCharon) This issue's original post is someone experiencing AccessDenied on Windows, and I'm not sure what they did to fix it. I don't know if your issue is related in any way. If it's unrelated, you should create a new issue if you don't find an existing one. I also suggest asking for help in one of the [community spaces](https://github.com/ziglang/zig/wiki/Community); you usually get quicker responses there.
> 
> Also I don't think you ever stating what version of zig you're using (the output of `zig version`). On a more recent build (`master` from https://ziglang.org/download/) the error might already be fixed, or you might encounter a different error.

Ok thanks, and my zig version is 0.13.0. And I'll create a issuse then.
```log
➜  Learn zig version
0.13.0
```

**Comment by KarryCharon** [comment]
> Regarding my issue, it was the build.zig Recompiling at every run. Deleting the cash folder fixed the issue but I am unsure what caused the behaviour in the first place.

Not the same with me...Caz even create a new project bug also
